### ISSUE TITLE
Attributes in include with limit/count

### ISSUE THREAD

**Issue by kollors** [sequelize/sequelize#7766]
## What you are doing?

```js
User.findAll({
  attributes: ['id', 'fullname'],
  include: [
    {
       attributes: ['id'],
       model: Tag, as: 'tags',
       offset: 3, limit: 2
     }
  ],
  limit: 2, offset: 1
}));
```

## What do you expect to happen?
I expect to get an association of users with tags, but if I remove the limit from tags everything is fine

## What is actually happening?
Cannot read property 'push' of undefined
TypeError: Cannot read property 'push' of undefined
    at Model.findAll.then.results (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/associations/has-many.js:230:66)
From previous event:
    at HasMany.get (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/associations/has-many.js:221:35)
    at Promise.map.include (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/model.js:1625:34)
From previous event:
    at Function._findSeparate (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/model.js:1600:20)
    at Promise.try.then.then.then.tap.then.results (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/model.js:1568:20)
    at runCallback (timers.js:800:20)
    at tryOnImmediate (timers.js:762:5)
    at processImmediate [as _immediateCallback] (timers.js:733:5)
From previous event:
    at Function.findAll (/home/niko/WebstormProjects/expressjs-seed/node_modules/sequelize/lib/model.js:1555:8)
    at router.get (/home/niko/WebstormProjects/expressjs-seed/routes/index.js:118:17)
    at Layer.handle [as handle_request] (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/layer.js:95:5)
    at next (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/route.js:137:13)
    at Route.dispatch (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/layer.js:95:5)
    at /home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:281:22
    at Function.process_params (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:335:12)
    at next (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:275:10)
    at Function.handle (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:174:3)
    at router (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:47:12)
    at Layer.handle [as handle_request] (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:317:13)
    at /home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:335:12)
    at next (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:275:10)
    at SessionStrategy.strategy.pass (/home/niko/WebstormProjects/expressjs-seed/node_modules/passport/lib/middleware/authenticate.js:325:9)
    at SessionStrategy.authenticate (/home/niko/WebstormProjects/expressjs-seed/node_modules/passport/lib/strategies/session.js:71:10)
    at attempt (/home/niko/WebstormProjects/expressjs-seed/node_modules/passport/lib/middleware/authenticate.js:348:16)
    at authenticate (/home/niko/WebstormProjects/expressjs-seed/node_modules/passport/lib/middleware/authenticate.js:349:7)
    at Layer.handle [as handle_request] (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/layer.js:95:5)
    at trim_prefix (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:317:13)
    at /home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:284:7
    at Function.process_params (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:335:12)
    at next (/home/niko/WebstormProjects/expressjs-seed/node_modules/express/lib/router/index.js:275:10)

__Dialect:__ mysql2
__Database version:__ 1.3.3
__Sequelize version:__ 4.0.0

**Comment by justintien** [comment]
```js
table1.findOne({
      include: [{
        as: 'table2',
        model: table2
        order: [
          ['create_time', 'desc']
        ],
        limit: 1
      }]
    });

2017-10-12T08:04:33.927Z sequelize:pool connection released
TypeError: Cannot read property 'push' of undefined
    at Model.findAll.then.results (/app/node_modules/sequelize/lib/associations/has-many.js:231:66)

if remove order & limit... work fine
```

**Comment by antonioasmoura** [comment]
facing same issue.

**Comment by narin** [comment]
Also facing the same issue

**Comment by rue-dhurd** [comment]
Since this is the top result in Google when searching for this, I'm going to post a clarification. Sequelize is in the right here, but it would be nice if the error message called out what was happening.

By introducing a JOIN and trying to limit the JOINed table's number of rows, we're forced to make two queries:

```js
User.findAll({
    where: { id: 1 },
    attributes: ['name'],
    include: [{
      model: Comment,
      attributes: ['text']
      limit: 1,
    }],
})
```
```sql
SELECT `users`.`name`
FROM `users` AS `users`
WHERE `users`.`id` = 1;

SELECT `comments`.`text`
FROM `comments` AS `comments`
WHERE `comments`.`user_id` IN (1) LIMIT 1;
```

Since the data isn't returned from the DB with `User` and `Comment` already associated, Sequelize has to do that on it's own. To make the connection, Sequelize needs the `user_id` attribute on `Comment` to determine which `User` it belongs to. But because we are specifying which attributes to select from `Comment` and not including the `user_id`, we've made it impossible to connect the dots. If we made the query with `attributes: ['user_id', 'text']` everything would work as expected.

---

TL;DR: Include the foreign key in your association table's `attributes`.

**Comment by kachalovsky** [comment]
Thank you @rue-dhurd. I quickly fixed this problem using your explanation!

**Comment by ThaisFaria** [comment]
@rue-dhurd Your answer makes sense but it has lead me to understang a secondary problem. I am doing this:

`    userModel.hasMany(applicationModel, {
      as: 'applications',
      foreignKey: { name: 'userId' },
    });

const models = await userModel
      .findAll({
        where,
        include: [{
            model: applicationModel,
            as: 'applications',
            required: false,
            where: {
              userId: { $col: 'user.id' },
            },
            order: [
              ['createdAt', 'DESC'],
            ],
            limit: 1,
        }],
      })`

**Comment by ThaisFaria** [comment]
@rue-dhurd Your answer makes sense but it has lead me to understang a secondary problem. I am doing this:

```    
userModel.hasMany(applicationModel, {
    as: 'applications',
    foreignKey: { name: 'userId' },
});

const models = await userModel
      .findAll({
        where,
        include: [{
            model: applicationModel,
            as: 'applications',
            required: false,
            where: {
              userId: { $col: 'user.id' },
            },
            order: [
              ['createdAt', 'DESC'],
            ],
            limit: 1,
        }],
      })
```

I believe this error is due to the quey being done seperately and having no access to the external result. I am not sure how to work around this without querying the user first and the application second. It works fine without the `limit 1`.
